rust/agama-network/src/nm/client.rs

```diff
@@ -0,0 +1,114 @@
+//! NetworkManager client.
+use super::dbus::{connection_from_dbus, connection_to_dbus, merge_dbus_connections};
+use super::model::NmDeviceType;
+use super::proxies::{ConnectionProxy, DeviceProxy, NetworkManagerProxy, SettingsProxy};
+use crate::model::{Connection, Device};
+use agama_lib::error::ServiceError;
+use uuid::Uuid;
+use zbus;
+
+/// Simplified NetworkManager D-Bus client.
+///
+/// Implements a minimal API to be used internally. At this point, it allows to query the list of
+/// network devices and connections, converting them to its own data types.
+pub struct NetworkManagerClient<'a> {
+    connection: zbus::Connection,
+    nm_proxy: NetworkManagerProxy<'a>,
+}
+
+impl<'a> NetworkManagerClient<'a> {
+    /// Creates a NetworkManagerClient connecting to the system bus.
+    pub async fn from_system() -> Result<NetworkManagerClient<'a>, ServiceError> {
+        let connection = zbus::Connection::system().await?;
+        Self::new(connection).await
+    }
+
+    /// Creates a NetworkManagerClient using the given D-Bus connection.
+    ///
+    /// * `connection`: D-Bus connection.
+    pub async fn new(
+        connection: zbus::Connection,
+    ) -> Result<NetworkManagerClient<'a>, ServiceError> {
+        Ok(Self {
+            nm_proxy: NetworkManagerProxy::new(&connection).await?,
+            connection,
+        })
+    }
+
+    /// Returns the list of network devices.
+    pub async fn devices(&self) -> Result<Vec<Device>, ServiceError> {
+        let mut devs = vec![];
+        for path in &self.nm_proxy.get_devices().await? {
+            let proxy = DeviceProxy::builder(&self.connection)
+                .path(path.as_str())?
+                .build()
+                .await?;
+
+            let device_type = NmDeviceType(proxy.device_type().await?);
+            if let Ok(device_type) = device_type.try_into() {
+                devs.push(Device {
+                    name: proxy.interface().await?,
+                    ty: device_type,
+                });
+            } else {
+                // TODO: use a logger
+                eprintln!("Unknown device type {:?}", &device_type);
+            }
+        }
+
+        Ok(devs)
+    }
+
+    /// Returns the list of network connections.
+    pub async fn connections(&self) -> Result<Vec<Connection>, ServiceError> {
+        let proxy = SettingsProxy::new(&self.connection).await?;
+        let paths = proxy.list_connections().await?;
+        let mut connections: Vec<Connection> = vec![];
+        for path in paths {
+            let proxy = ConnectionProxy::builder(&self.connection)
```

Comment:
NP: I just noticed it here, but having connection in NM and dbus connection using same name can cause confusion.

Reply:
> 

Yes, you are right, and I did not want to use `ConnectionIface` or something like that (it is redundant). I would keep it as it is and we can improve it later.

Reply:
Yes, you are right, and I did not want to use ConnectionIface or something like that (it is redundant). I would keep it as it is and we can improve it later.